bpf/libbpf/libbpf.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package libbpf
+
+import (
+	"fmt"
+	"unsafe"
+
+	"github.com/projectcalico/felix/bpf"
+)
+
+// #include "libbpf_api.h"
+import "C"
+
+type Obj struct {
+	obj *C.struct_bpf_object
+}
+
+type TCOpts struct {
+	opts C.struct_bpf_tc_opts
+}
+
+type Map struct {
+	bpfMap *C.struct_bpf_map
+	bpfObj *C.struct_bpf_object
+}
+
+const MapTypeProgrArray = C.BPF_MAP_TYPE_PROG_ARRAY
+
+type QdiskHook string
+
+const (
+	QdiskIngress QdiskHook = "ingress"
+	QdiskEgress  QdiskHook = "egress"
+)
+
+func (m *Map) Name() string {
+	name, err := C.bpf_map__name(m.bpfMap)
+	if err != nil {
+		return ""
+	}
+	return C.GoString(name)
+}
+
+func (m *Map) Type() int {
+	mapType, err := C.bpf_map__type(m.bpfMap)
+	if err != nil {
+		return -1
+	}
+	return int(mapType)
+}
+
+func (m *Map) SetPinPath(path string) error {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+	err := C.bpf_map__set_pin_path(m.bpfMap, cPath)
```

Comment:
I think this error handling will work, comparing with the libbpf code.  It looks like it returns `-errno` on failure.